### PR TITLE
Feat/create new table dialog box

### LINKED ISSUE #75: Build html "create new table" dialogue overlay for data import page
After clicking "Create New Table" (page created with the closing of #74), a dialogue overlay should appear where the user adds table name and description. 

Ensure that we are building the finalised design (check that #70 is closed)

Image

### LINKED ISSUE #85: Function to run check that "create new table" name doesn't already exist

### LINKED ISSUE #128: Fix: Fly.io deployment fails with error about "CSRF_TOKEN"

### PR DESCRIPTION
**Pull request by signekb**
## Description

- This PR adds a dialog overlay on the data import page with two fields: `name` and `description`
- The `name` field includes two validators:
    - A validator that checks that the table name doesn't include special characters (only a-z, A-Z, 0-9, hyphen (-), and/or underscore (_)
    - A validator that checks that a table with the same name doesn't exist already
- If table `name` and `description` passes validation, the information is saved in the `TableMetadata`
-  I have included tests for the validators, the form, and the data import view
- In addition, this PR adds a (?) button on the data import page that opens a dialog where we can write a helper message and a link to our documentation. (this needs to be updated later ofc)

I have added a detailed description of the changes in the bottom of this PR to (hopefully) ease the review process, if something is unclear. 

## Related Issues

Closes #75, #76 and #125

## Testing

- [X] Yes

## Reviewer Focus
 

- Anything in the code that's unclear? / Suggestions for better ways of doing this? 
- Variable naming and doc string descriptions
- Tests: Enough / too many / some that aren't necessary?

## Checklist

For all PRs that are not general documentation

- [X] Tests accompany or reflect changes to the code
- [X] Tests ran and passed locally
- [X] Ran the linter and formatter (NB: But only on the files relevant to this PR)
- [X] Build has passed locally
- [X] Relevant documentation has been updated

### Overview of changes
Here's an overview of the changes in the order they appear in the directory (I hope this will ease the review process): 

- **app/forms.py**: 
    - I have created a `forms.py` file with a [ModelForm](https://docs.djangoproject.com/en/5.0/topics/forms/modelforms/) called `TableMetadataForm` . This class uses the model `TabelMetadata` and the fields `name` and `description` from that model. It's these fields that we want users to fill out in the dialog overlay when they create a table (as shown in the wireframes on Figma). 
    - Within this form, I call the two validators, I have created in `validators.py`
- **app/templates/data-import.html**:
    -  I have created a "create-table-dialog" that opens, when a user clicks "create new table". This dialog includes two validations, which will be described below.
    - In addition, I have added a help button that opens the "help-dialog" when clicked.
- **app/validators/field_validators.py**:
    - `validate_no_special_characters`: A regex validator that fails, if the field includes special characters. We can reuse this validators for other fields where we don't want special characters
    - `validate_table_name_does_not_exist`: A validator that checks that a table name doesn't exist in the database already.
- **app/views.py**
    - I have updated the `data_import` view to include the `TableMetadataForm`. When the dialog opens, the request will be `GET` and a blank form is created. When the user clicks "Create", the request changes to `POST` (see `data-import.html`) and the form's content is checked by the validators. If the validation fails, the error message from the relevant validator is shown. If the validation succeeds, it's saved and the url is redirected to `file-upload/{table_id}`, where {table-id} is the id of the table that was just created. This is useful for the file-upload page, so this page can get the table name from the id.
- **app/tests/tests_forms.py**: 
    - Tests of the form `TableMetadataForm`. 
- **app/tests/tests_validators.py**: 
    - Tests for the two validators `validate_no_special_characters` and `validate_table_name_does_not_exist`
- **app/tests/tests_views.py**: 
    - Tests for the view (does it render, redirect, etc.) when we expect it to

### CODE REVIEW

**Comment by signekb** [comment]
@lwjohnst86 @philter87, in the fly preview, it fails whenever the request goes to "POST" (I think). I'm not sure why this happens. Do you have an idea?

**Review by philter87 — DISMISSED** [review]
Beautiful :) I have much to learn about making a clean PR! I think everything is very clear. 

I don't think testing the view is so important, because you have tested the validators and the form.

**Comment by philter87** [comment]
> @lwjohnst86 @philter87, in the fly preview, it fails whenever the request goes to "POST" (I think). I'm not sure why this happens. Do you have an idea?

I think this is weird too! I didn't do for my hobby project.

However, I think it should work if you add: 
CSRF_TRUSTED_ORIGINS = ["https://seedcase-sprout*.fly.dev"]

But I might need to investigate a bit more to know if it is secure

File: app/forms.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    validate_table_name_does_not_exist,
+)
+
+
+class TableMetadataForm(forms.ModelForm):
+    class Meta:
+        model = TableMetadata
+        fields = ["name", "description"]
+
+    def clean_name(self):
```
[the remark this answers is not in the document]
Comment (signekb) [reply app/forms.py:15]: Great idea. Done! 🎉

**Comment by signekb** [comment]
I'm through my edits based on feedback from @philter87, and I have now added tests for the view

**Review by philter87 — DISMISSED** [review]

**Comment by philter87** [comment]
I did a couple of things to fix the CSRF POST request issue:
- I added CSRF_TRUSTED_ORIGINS = [f"https://{APP_NAME}.fly.dev/"]
- I resolved the merge conflict, because it prevented deploys to fly.io (needed to validate the CSRF issue)

File: app/tests/tests_validators.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        # Act
+        with self.assertRaises(ValidationError) as context:
+            validate_no_special_characters(
+                field_name=field_name, field_value=field_value
+            )
+
+        # Assert
+        self.assertEqual(
+            context.exception.message,
+            f"Please provide a {field_name} without special characters",
```
[the remark this answers is not in the document]
Comment (signekb) [reply app/tests/tests_validators.py:30]: Uh, I like that! Will change it

File: app/templates/data-import.html [hunk 3]
```diff
@@ -4,13 +4,11 @@
...
     <div class="row center-align">
       <div>
-        <a href="create-new-table">
-          <article class="card-button-size">
-            <!-- In beercss, a small material icon is added with <i class="large">add</i>,
-             but to create a large icon, I grabbed the svg from https://fonts.google.com/icons-->
-            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 -960 960 960" ><path d="M440-440H200v-80h240v-240h80v240h240v80H520v240h-80v-240Z"/></svg>
-          </article>
-        </a>
+        <article class="card-button-size" data-ui="#create-table-dialog">
```
Comment (philter87) [inline app/templates/data-import.html:7]: A small thing. The cursor is not turning into a hand when I hover the button - this disappeared when the <a> tag was removed. Add "cursor: pointer" to "card-button-size" or add style="cursor:pointer"

**Review by philter87 — DISMISSED** [review]

File: app/templates/data-import.html (hunk 3 above)
Comment (signekb) [reply app/templates/data-import.html:7]: Aha! Good catch! I will add that :hand:

**Review by philter87 — DISMISSED** [review]

**Review by lwjohnst86 — APPROVED** [review]
Looks so fantastic! Great job 😁 I only made some minor changes and added a test to check for invalid input.

**Comment by signekb** [comment]
@lwjohnst86 
I believe you have created a duplicate test? Isn't the invalid input you added on L61-65 the same test as the one on L8-15?
https://github.com/seedcase-project/seedcase-sprout/blob/a9fe5ec8e1c73bc18376b29abdf6c4949a4933a7/app/tests/tests_forms.py#L8-L15

https://github.com/seedcase-project/seedcase-sprout/blob/a9fe5ec8e1c73bc18376b29abdf6c4949a4933a7/app/tests/tests_forms.py#L61-L65

**Comment by lwjohnst86** [comment]
Ah, you're right @signekb, oops!

**Comment by signekb** [comment]
I guess that's the danger of skipping reviews ;)

**Comment by lwjohnst86** [comment]
😆 that's very true!

**Pull request merged**